hnix-store-core/src/System/Nix/Internal/Nar/Streamer.hs

```diff
@@ -19,22 +22,41 @@ import           System.FilePath                  ( (</>) )
 import qualified System.Nix.Internal.Nar.Effects as Nar
 
 
+-- | NarSource 
+-- the source to provide nar to the handler
+type NarSource m =  (ByteString -> m ()) -> m ()
+
+
+-- | dumpString
+-- dump a string to nar
+dumpString :: forall m . IO.MonadIO m => ByteString -> NarSource m
+dumpString text yield = do
```

Comment:
This is strange. Even confusing to read.

Type signature has 1 argument, but the function has 2.

I'd argumented that it is more confusing to read than my style. Because if more of such types are used - that would make type signatures unreadable & so not useful for programmers.

Reply:
Agree that it is confusing to read, especially with remote part getting stashed.
Let me justify the this style.

It is done in CPS style. Under Yoneda Lemma in category theory, `NarSource m` is isomorphic to `ByteString`.
It is to align with code already in the codebase.
```haskell
runOpArgsIO
  :: WorkerOp
  -> ((Data.ByteString.ByteString -> MonadStore ()) -> MonadStore ())
  -> MonadStore ()`. 
```
Using this style instead of directly `ByteString`, I think is it to enable to chunks the `IO` input instead of feeding the whole nar at once to achieve better performance. 

It would be more reasonable once we look at the API of addToStore. To generify the the original API to more than taking in a `FilePath`, we need the `NarSource` to take a handler `ByteString -> m ()` to actually process the nar bytes.

Reply:
The Remote Part is something like this:

```haskell
-- | Pack `Nar` and add it to the store.
addToStore1
  :: StorePathName        -- ^ Name part of the newly created `StorePath`
  -> NarSource MonadStore -- ^ provide nar stream
  -> Bool                 -- ^ Add target directory recursively
  -> (FilePath -> Bool)   -- ^ Path filter function
  -> RepairFlag           -- ^ Only used by local store backend
  -> MonadStore StorePath
addToStore1 name source recursive _pathFilter _repair = do
  runOpArgsIO AddToStore $ \yield -> do
    yield $ toStrict $ runPut $ do
      putText $ unStorePathName name
      putBool $ not $ algoName @Hash.SHA256 == "sha256" && recursive
      putBool recursive
      putText $ System.Nix.Hash.algoName @Hash.SHA256
    source yield
  sockGetPath
```

Reply:
In such cases, you can just open another PR, which depends on this branch & mention it depends here & thankfully GitHub allows to mark the PR as a draft.

Having that one additional level of nesting is perfectly reasonable.

Reply:
Sure the full version would be in draft #179